sway/commands/move.c

```diff
@@ -316,72 +306,88 @@ static bool container_move_in_direction(struct sway_container *container,
+				container->height_fraction = container->width_fraction = 0;
+				workspace_update_representation(current->workspace);
+				wrapped = true;
+			} else {
+				// Keep looking for a parallel parent
+				current = current->parent;
 			}
+			continue;
 		}
 
-		current = current->parent;
-	}
+		// Only scratchpad hidden containers don't have siblings
+		// so siblings != NULL here
+		list_t *siblings = container_get_siblings(current);
+		int index = list_find(siblings, current);
+		int desired = index + offs;
+		struct sway_container *target =
+			desired == -1 || desired == siblings->length ?
+			NULL : siblings->items[desired];
+
+		if (current == container) {
+			if (target) {
+				// Container will swap with or descend into its neighbor
+				container_move_to_container_from_direction(container,
+						target, move_dir);
+				return true;
+			} else if (!container->parent) {
+				// Container is at workspace level so we move it to the
+				// next workspace if possible
+				return container_move_to_next_output(container,
+						current->workspace->output, move_dir);
+			} else {
+				// Container has escaped its immediate parallel parent
+				current = current->parent;
+				continue;
+			}
+		}
 
-	// Maybe rejigger the workspace
-	struct sway_workspace *ws = container->workspace;
-	if (ws) {
-		if (!is_parallel(ws->layout, move_dir)) {
-			workspace_rejigger(ws, container, move_dir);
+		// Loop is ending
+		if (target) {
+			// Container will move in with its cousin
+			container_move_to_container_from_direction(container,
+					target, move_dir);
 			return true;
-		} else if (ws->layout == L_TABBED || ws->layout == L_STACKED) {
-			workspace_rejigger(ws, container, move_dir);
+		} else if (!wrapped && container->parent->workspace &&
+				container->parent->children->length == 1) {
+			// Treat singlton children as if they are at workspace level like i3
+			// https://github.com/i3/i3/blob/1d9160f2d247dbaa83fb62f02fd7041dec767fc2/src/move.c#L367
+			return container_move_to_next_output(container,
+					current->workspace->output, move_dir);
+		} else if (current->parent) {
+			// Container will move in with its parent
+			container_insert_child(current->parent, container,
+					index + (offs < 0 ? 0 : 1));
 			return true;
-		}
-
-		// Try adjacent output
-		struct sway_output *output =
-			output_get_in_direction(container->workspace->output, move_dir);
-		if (output) {
-			struct sway_workspace *ws = output_get_active_workspace(output);
-			if (!sway_assert(ws, "Expected output to have a workspace")) {
-				return false;
-			}
-			container_move_to_workspace_from_direction(container, ws, move_dir);
+		} else {
+			// Container will move to workspace level,
+			// may be re-split by workspace_layout
+			workspace_insert_tiling(current->workspace, container,
+					index + (offs < 0 ? 0 : 1));
 			return true;
 		}
-		sway_log(SWAY_DEBUG, "Hit edge of output, nowhere else to go");
 	}
+	// Unreachable
 	return false;
```

Comment:
So I added that there because the compiler gets upset that it reaches the end of the function without a return, and it does that because it doesn't know that since I changed the loop, "current" will actually never be null.

So the real problem here is I silenced the compiler warning without correcting the code. The loop condition should actually just be `while(true)` – the behavior is the same but it doesn't mislead readers, and the compiler, to think there is an invariant there.

Instead I think I'll move the declarations from the middle of the loop to before the loop, and the if block at the end to outside the loop. Once it gets to that final block it can't repeat anyway since every branch of the if has a return.

Reply:
Actually better yet I can just introduce a variable to be the loop invariant.